Toolhouse.Monitoring/HttpRequestMetrics.cs

```diff
@@ -0,0 +1,148 @@
+        public static T Instrument<T>(string name, Func<Labels, T> request)
+        {
+            return InstrumentWithMetrics(name, requestMetrics =>
+            {
+                return requestMetrics.PerformRequest(() =>
+                {
+                    return request(requestMetrics.m_labels);
+                });
+            });
+        }
+
+        private HttpRequestMetrics(string name)
+        {
+            m_labels = new Labels();
+            m_name = name;
+            m_stopwatch = new Stopwatch();
+            m_currentRequests = Prometheus.Metrics.CreateGauge(string.Format("{0}_current_requests", m_name), string.Empty, m_labels.BackendLabelKey);
+        }
+
+        private static void InstrumentWithMetrics(string name, Action<HttpRequestMetrics> instrumentWithMetrics)
+        {
+            InstrumentWithMetrics(name, requestMetrics =>
+            {
+                instrumentWithMetrics(requestMetrics);
+                return true;
+            });
+        }
+
+        private static T InstrumentWithMetrics<T>(string name, Func<HttpRequestMetrics, T> instrumentWithMetrics)
+        {
+            if (name.Any(char.IsWhiteSpace))
+            {
+                throw new ArgumentException("Parameter must be a single word.", "name");
+            }
+
+            HttpRequestMetrics requestMetrics = new HttpRequestMetrics(name);
+            return instrumentWithMetrics(requestMetrics);
+        }
+
+        private void PerformRequest(Action request)
+        {
+            PerformRequest(() =>
+            {
+                request();
+                return true;
+            });
+        }
+
+        private T PerformRequest<T>(Func<T> request)
+        {
+            StartGatheringMetrics();
+            T result = request();
+            StopGatheringMetrics();
+            RecordMetrics();
+            return result;
+        }
+
+        private void StartGatheringMetrics()
+        {
+            Prometheus.Metrics.CreateCounter(string.Format("{0}_requests_total", m_name), string.Empty, m_labels.BackendLabelKey)
+                .Labels(m_labels[m_labels.BackendLabelKey]).Inc();
+            m_currentRequests.Labels(m_labels[m_labels.BackendLabelKey]).Inc();
+            m_stopwatch.Start();
+        }
+
+        private void StopGatheringMetrics()
+        {
+            m_stopwatch.Stop();
+            m_currentRequests.Labels(m_labels[m_labels.BackendLabelKey]).Dec();
+        }
+
+        private void RecordMetrics()
+        {
+            Prometheus.Metrics.CreateCounter(string.Format("{0}_responses_total", m_name), string.Empty, m_labels.Keys)
+                .Labels(m_labels.Values).Inc();
+            Prometheus.Metrics.CreateHistogram(string.Format("{0}_request_duration_seconds", m_name), string.Empty, null, m_labels.Keys)
+                .Labels(m_labels.Values).Observe(m_stopwatch.Elapsed.TotalSeconds);
+        }
+
+        readonly Labels m_labels;
```

Comment:
I'd like the layout of new classes to match layout of existing classes in this project. Please:

- Move fields to the top of classes
- Omit the `m_` prefix
- Ensure all fields have an appropriate access modifier (`private`, `protected`, etc.)

(Ideally we'd have some automated checking of this, but I haven't had the time to set it up here)

Reply:
I agree with the second point as long as there is some way to distinguish private member variables from instance variables and properties. Maybe we should have a more in-depth discussion about this.